src/uri/port.rs

```diff
@@ -0,0 +1,64 @@
+use std::{fmt, num};
+use std::str::FromStr;
+
+/// The port component of a URI.
+#[derive(Debug)]
+pub struct Port<'a> {
+    bytes: &'a str,
+    port: u16,
+}
+
+impl<'a> Port<'a> {
+    /// Converts a `str` to a port number.
+    ///
+    /// The supplied `str` must be a valid u16.
+    pub(crate) fn from_str(bytes: &'a str) -> Result<Self, num::ParseIntError> {
+        u16::from_str(bytes).and_then(|port| Ok(Port { port, bytes }))
```

Comment:
The BNF of a URI port is `*DIGIT`, see https://tools.ietf.org/html/rfc3986#section-3.2.3. However `from_str` allows an optional `+` sign, see https://doc.rust-lang.org/nightly/std/primitive.u16.html#method.from_str_radix.

Reply:
In theory, this should already have been validated when constructing the `Uri`, so checking again here is not necessary.